### PR TITLE
PR: Send comm config again if comm times out

### LINKED ISSUE #11137: TimeoutError when running cells after a kernel restart
Full traceback below; cell runs fine via highlight + "Run selection or current line"; behavior's fixed upon restarting kernel, but returns randomly after finishing a TensorFlow training session (didn't test in other settings).

```python
Traceback (most recent call last):

  File "D:\Anaconda\envs\s4_env\lib\site-packages\spyder_kernels\comms\commbase.py", line 542, in __call__
    call_dict, call_data, self._comm_id)
  File "D:\Anaconda\envs\s4_env\lib\site-packages\spyder_kernels\comms\commbase.py", line 415, in _get_call_return_value
    self._wait_reply(call_id, call_name, timeout)
  File "D:\Anaconda\envs\s4_env\lib\site-packages\spyder_kernels\comms\frontendcomm.py", line 161, in _wait_reply
    call_name))

TimeoutError: Timeout while waiting for 'run_cell' reply
```

<hr>

**Update**: reproduced immediately after restarting kernel and working with basic numpy operations

<hr>

**Update 2**: interrupting frozen run-cell command yields below:

```python
Traceback (most recent call last):

  File "<ipython-input-4-633e420a5051>", line 1, in <module>
    runcell('[], #5', 'C:/DL_code/dev_ensemble_subset.py')
  File "D:\Anaconda\envs\s4_env\lib\site-packages\spyder_kernels\customize\spydercustomize.py", line 539, in runcell
    cell_code = frontend_request().run_cell(cellname, filename)
  File "D:\Anaconda\envs\s4_env\lib\site-packages\spyder_kernels\comms\commbase.py", line 542, in __call__
    call_dict, call_data, self._comm_id)
  File "D:\Anaconda\envs\s4_env\lib\site-packages\spyder_kernels\comms\commbase.py", line 415, in _get_call_return_value
    self._wait_reply(call_id, call_name, timeout)
  File "D:\Anaconda\envs\s4_env\lib\site-packages\spyder_kernels\comms\frontendcomm.py", line 167, in _wait_reply
    time.sleep(0.01)

KeyboardInterrupt
```

### LINKED ISSUE #11406: Kernel dies silently and doesn't restart properly.
### Issue Report Checklist

* [ ] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [ ] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [ ] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [ ] Could not reproduce inside ``jupyter qtconsole`` (if console-related)
* [ ] Tried basic troubleshooting (if a bug/error)
    * [ ] Restarted Spyder
    * [ ] Reset preferences with ``spyder --reset``
    * [ ] Reinstalled the latest version of [Anaconda](https://www.anaconda.com/download/)
    * [ ] Tried the other applicable steps from the Troubleshooting Guide
* [ ] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below

### What steps reproduce the problem?

1. Type in the kernel:
```
import numpy as np
a = np.zeros((2, 18030, 162542200), float) * np.nan
```
2. Wait for python to crash
3. When you get a prompt back, try running a cell.

### What is the expected output? What do you see instead?

spyder:
```
Python 3.7.5 (default, Nov  1 2019, 02:16:23) 
Type "copyright", "credits" or "license" for more information.

IPython 7.10.2 -- An enhanced Interactive Python.

In [1]: import numpy as np
   ...: a = np.zeros((2, 18030, 162542200), float) * np.nan

In [1]: 
```
qtconsole:
```
Jupyter QtConsole 4.7.0.dev0
Python 3.7.5 (default, Nov  1 2019, 02:16:23) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.10.2 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import numpy as np
   ...: a = np.zeros((2, 18030, 162542200), float) * np.nan

Kernel died, restarting


 
Kernel died, restarting


 
Kernel died, restarting


 

In [1]: 

In [1]: 
```

python3:
```
Python 3.7.5 (default, Nov  1 2019, 02:16:23) 
[Clang 11.0.0 (clang-1100.0.33.8)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> a = np.zeros((2, 18030, 162542200), float) * np.nan
zsh: killed     python3
```

After the prompt returns and I run a cell, I get:
```
In [1]: runcell(0, 'test.py')
This command failed to be executed because an error occurred while trying to get the cell code from Spyder's editor. The error was:


An exception has occurred, use %tb to see the full traceback.

CommError: Can't make a request to a closed comm
```
### Paste Traceback/Error Below (if applicable)

```python-traceback

PASTE TRACEBACK HERE

```

## Versions

* Spyder version: 
* Python version: 
* Qt version: 
* PyQt version: 
* Operating System name/version: 

### Dependencies

```
PASTE DEPENDENCIES HERE
```

### PR DESCRIPTION
**Pull request by impact27**
## Description of Changes
* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

If the comm config was not received, the kernel tries to send it again before raising a TimeoutError

It also restart the kernel in a thread to avoid freezing the frontend and shows a `loading_page`.
This also fixes a bug where the loading page might not be hidden if the kernel starts too quickly.

### Issue(s) Resolved

Fixes #11137 
Fixes #11406
Depends on spyder-ide/spyder-kernels#192

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: impact27

### CODE REVIEW

File: spyder/plugins/ipythonconsole/comms/kernelcomm.py [hunk 1]
```diff
@@ -108,40 +123,45 @@ def _get_call_return_value(self, call_dict, call_data, comm_id):
...
 
     def _wait_reply(self, call_id, call_name, timeout):
         """Wait for the other side reply."""
-        if call_id in self._reply_inbox:
+
+        def got_reply():
+            return call_id in self._reply_inbox
+
+        timeout_msg = "Timeout while waiting for {}".format(
+                            self._reply_waitlist)
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/comms/kernelcomm.py]: This is incorrectly formatted. Please change it to 

```python
timeout_msg = "Timeout while waiting for {}".format(
    self._reply_waitlist)
```

File: spyder/plugins/ipythonconsole/comms/kernelcomm.py [hunk 2]
```diff
@@ -108,40 +123,45 @@ def _get_call_return_value(self, call_dict, call_data, comm_id):
...
+        self._wait(got_reply, self._sig_got_reply, timeout_msg, timeout)
+
+    def _wait(self, condition, signal, timeout_msg, timeout):
+        """Wait for a condition"""
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/comms/kernelcomm.py]: Please expand this docstring to describe what `condition`, `signal`, `timeout_msg` and `timeout` are.

File: spyder/plugins/ipythonconsole/comms/tests/test_comms.py [hunk 3]
```diff
@@ -102,8 +102,14 @@ def comms(kernel):
     frontend_comm = FrontendComm(kernel)
     kernel_comm = KernelComm()
 
+    def dummy(port):
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/comms/tests/test_comms.py]: Please change the name of this function to something more meaningful. I mean, I don't understand what this `dummy` is doing below.

File: spyder/plugins/ipythonconsole/widgets/client.py [hunk 4]
```diff
@@ -227,8 +227,8 @@ def remove_stderr_file(self):
         except Exception:
             pass
 
-    def configure_shellwidget(self, give_focus=True):
-        """Configure shellwidget after kernel is started"""
+    def post_configure_shellwidget(self, give_focus=True):
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/client.py]: Please leave this method's name as it was.

File: spyder/plugins/ipythonconsole/widgets/client.py [hunk 5]
```diff
@@ -273,12 +273,20 @@ def configure_shellwidget(self, give_focus=True):
         # To apply style
         self.set_color_scheme(self.shellwidget.syntax_style, reset=False)
 
-        # To hide the loading page
-        self.shellwidget.sig_prompt_ready.connect(self._hide_loading_page)
+    def pre_configure_shellwidget(self):
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/client.py]: There's no configuration done to the shellwidget in this method. Therefore, I  think it's better to name it `before_prompt_is_ready`.

File: spyder/plugins/ipythonconsole/widgets/client.py [hunk 6]
```diff
@@ -273,12 +273,20 @@ def configure_shellwidget(self, give_focus=True):
...
 
-        # To hide the loading page
-        self.shellwidget.sig_prompt_ready.connect(self._hide_loading_page)
+    def pre_configure_shellwidget(self):
+        """Configure shellwidget before kernel is connected."""
+        self._show_loading_page()
+        self.shellwidget.sig_prompt_ready.connect(
+            self.post_prompt_configure_shellwidget)
 
+    def post_prompt_configure_shellwidget(self):
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/client.py]: Please name this method `when_prompt_is_ready`.

File: spyder/plugins/ipythonconsole/widgets/client.py [hunk 7]
```diff
@@ -273,12 +273,20 @@ def configure_shellwidget(self, give_focus=True):
...
+        """Configuration after the prompt is shown."""
+        # To hide the loading page
+        self._hide_loading_page()
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/client.py]: Please add a blank line below this one.

File: spyder/plugins/ipythonconsole/widgets/client.py [hunk 8]
```diff
@@ -535,32 +543,51 @@ def restart_kernel(self):
...
+        try:
+            self.shellwidget.kernel_manager.restart_kernel(
+                stderr=self.stderr_handle)
+        except RuntimeError as e:
+            self.restart_thread.error = e
+
+    def _restart_thread_end(self):
+        """Finishes the restarting of the kernel."""
+        sw = self.shellwidget
+        e = self.restart_thread.error
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/client.py]: `e` is not descriptive enough. Please change this to `error`.

File: spyder/plugins/ipythonconsole/widgets/client.py [hunk 9]
```diff
@@ -535,32 +543,51 @@ def restart_kernel(self):
...
+        if e is not None:
+            sw._append_plain_text(
+                _('Error restarting kernel: %s\n') % e,
+                before_prompt=True
+            )
+        else:
+            # Reopen comm
+            sw.spyder_kernel_comm.open_comm(sw.kernel_client)
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/client.py]: Please add a blank line below this one.

File: spyder/plugins/ipythonconsole/comms/tests/test_comms.py (hunk 3 above)
Comment (impact27) [reply spyder/plugins/ipythonconsole/comms/tests/test_comms.py]: This function is not doing anything. I don't use real comms in this test so it would fail if I try to set a port.

Comment (ccordoba12) [reply spyder/plugins/ipythonconsole/comms/tests/test_comms.py]: Yeah, but I don't understand it's purpose. Could you give it a better name so it's clearer why it's necessary here?

Comment (impact27) [reply spyder/plugins/ipythonconsole/comms/tests/test_comms.py]: done

File: continuous_integration/azure/install.bat [hunk 10]
```diff
@@ -38,7 +38,7 @@ conda create -n jedi-test-env -q -y python=3.6 loghub spyder-kernels -c spyder-i
 if errorlevel 1 exit 1
 
 :: Install spyder-kernels from master
-pip install -q --no-deps git+https://github.com/spyder-ide/spyder-kernels
+pip install -q --no-deps git+https://github.com/spyder-ide/spyder-kernels.git@refs/pull/192/head
```
Comment (ccordoba12) [inline continuous_integration/azure/install.bat]: Please revert these changes to our installation scripts because I merged this PR.

File: spyder/plugins/ipythonconsole/comms/kernelcomm.py [hunk 11]
```diff
@@ -62,12 +65,27 @@ def shutdown_comm_channel(self):
...
-        if not self.kernel_client.comm_channel:
+        if queue_message:
+            # Send without comm_channel
             yield
             return
+
+        if not self.comm_channel_connected():
+            # Ask again for comm config
+            self.remote_call()._send_comm_config()
+            timeout = 10
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/comms/kernelcomm.py]: Is this 10 seconds? If that's the case, why do it need to be that big?

File: spyder/plugins/ipythonconsole/widgets/client.py [hunk 12]
```diff
@@ -144,7 +144,7 @@ def __init__(self, plugin, id_,
         # To keep a reference to the page to be displayed
         # in infowidget
         self.info_page = None
-        self._show_loading_page()
+        self.before_prompt_is_ready()
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/client.py]: I don't think this needs to be exposed publicly, so please change its name to `_before_prompt_is_ready`

File: spyder/plugins/ipythonconsole/comms/kernelcomm.py (hunk 11 above)
Comment (impact27) [reply spyder/plugins/ipythonconsole/comms/kernelcomm.py]: I have to wait for the kernel to launch and connect, which might take a while. 1s is too short. Also this is something that is only waited on once.

Comment (ccordoba12) [reply spyder/plugins/ipythonconsole/comms/kernelcomm.py]: Ok, then I have another question: what happens if the kernel takes more than 10 secs to launch? That's not an atypical case on Windows (I've seen cases where it takes 30 or 40 secs, and it could take even more after we merge PR #11327).

Comment (ccordoba12) [reply spyder/plugins/ipythonconsole/comms/kernelcomm.py]: Ok, then I have another question: what happens if the kernel takes more than 10 secs to launch? That's not an atypical case on Windows (I've seen cases where it takes 30 or 40 secs, and it could take even more after we merge PR #11327).

Comment (impact27) [reply spyder/plugins/ipythonconsole/comms/kernelcomm.py]: Then the function that is making the call would fail.

Comment (impact27) [reply spyder/plugins/ipythonconsole/comms/kernelcomm.py]: But the next function would wait again

Comment (impact27) [reply spyder/plugins/ipythonconsole/comms/kernelcomm.py]: I tried adding `time.sleep(15)` in `start.py::main()` and nothing bad happened.
